File: EliteFramework/EliteLayer.m
//
//  EliteLayer.m
//  EliteFramework
//
//  Created by Lifelong-Study on 2016/1/9.
//  Copyright © 2016年 Lifelong-Study. All rights reserved.
//

#import "EliteLayer.h"

@implementation CALayer (EliteLayerCategory)

// 渲染漸層效果
- (void)renderGradientEffectsWithDirection:(EliteDirection)direction colorWithCGColorsArray:(NSArray *)colorsArray
{
    // 初始化漸層效果
    CAGradientLayer *gradientLayer = [CAGradientLayer layer];
    gradientLayer.startPoint = [self getStartPointWithDirection:direction];
    gradientLayer.endPoint   = [self getEndPointWithDirection:direction];
    gradientLayer.frame      = [self bounds];
    gradientLayer.colors     = colorsArray;
    gradientLayer.name       = @"GradientEffectsLayer";
    
    // 套用漸層效果
    if ([[[[self sublayers] objectAtIndex:0] name] isEqualToString:@"GradientEffectsLayer"]){
        [self replaceSublayer:[[self sublayers] objectAtIndex:0] with:gradientLayer];
    } else {
        [self insertSublayer:gradientLayer atIndex:0];
    }
}

- (void)renderGradientEffectsWithDirection:(EliteDirection)direction colorWithUIColor:(UIColor *)colors, ...
{
    UIColor *color = colors;
    
    NSMutableArray *colorsArray = [[NSMutableArray alloc] init];
    
    va_list args;
    va_start(args, colors);
    
    [colorsArray addObject:(id)color.CGColor];
    while ((color = va_arg(args, UIColor *)) != nil) {
        [colorsArray addObject:(id)color.CGColor];
    }
    
    va_end(args);
    
    if (colorsArray.count > 0) {
        [self renderGradientEffectsWithDirection:direction colorWithCGColorsArray:colorsArray];
    }
}

- (void)renderGradientEffectsWithDirection:(EliteDirection)direction colorWithCGColor:(id)colors, ...
{
    id color = nil;
    
    NSMutableArray *colorsArray = [[NSMutableArray alloc] init];
    
    va_list args;
    va_start(args, colors);
    
    [colorsArray addObject:colors];
    while ((color = va_arg(args, id)) != nil) {
        [colorsArray addObject:color];
    }
    
    va_end(args);
    
    if (colorsArray.count > 0) {
        [self renderGradientEffectsWithDirection:direction colorWithCGColorsArray:colorsArray];
    }
}

//- (void)renderGradientEffectsWithDirection:(EliteDirection)direction colorWithCGColors:(NSArray *)colors
//{
//    id color = nil;
//    
//    NSMutableArray *colorsArray = [[NSMutableArray alloc] init];
//    
////    va_list args;
////    va_start(args, colors);
//    
//    [colorsArray addObject:colors];
//    while ((color = va_arg(args, id)) != nil) {
//        [colorsArray addObject:color];
//    }
//    
//    va_end(args);
//    
//    if (colorsArray.count > 0) {
//        [self renderGradientEffectsWithDirection:direction colorWithCGColorsArray:colorsArray];
//    }
//}

-  (CGPoint)getStartPointWithDirection:(EliteDirection)direction
{
    switch (direction) {
        case EliteDirectionFromTopToBottom:     return CGPointMake(0.5, 0);
        case EliteDirectionFromBottomToTop:     return CGPointMake(0.5, 1);
        case EliteDirectionFromLeftToRight:     return CGPointMake(0, 0.5);
        case EliteDirectionFromRigthToLeft:     return CGPointMake(1, 0.5);
        default:                                return CGPointMake(0.5, 0);
    }
}

-  (CGPoint)getEndPointWithDirection:(EliteDirection)direction
{
    switch (direction) {
        case EliteDirectionFromTopToBottom:     return CGPointMake(0.5, 1);
        case EliteDirectionFromBottomToTop:     return CGPointMake(0.5, 0);
        case EliteDirectionFromLeftToRight:     return CGPointMake(1, 0.5);
        case EliteDirectionFromRigthToLeft:     return CGPointMake(0, 0.5);
        default:                                return CGPointMake(0.5, 1);
    }
}

@end

File: EliteFramework/EliteLayer.swift
//
//  EliteLayer.swift
//  Transformer
//
//  Created by Lifelong-Study on 2016/1/19.
//  Copyright © 2016年 Lifelong-Study. All rights reserved.
//

import UIKit

//public enum EliteDirection {
//    case fromTopToBottom  // Top    -> Bottom
//    case fromBottomToTop  // Bottom -> Top
//    case fromLeftToRight  // Left   -> Right
//    case fromRightToLeft  // Right  -> Left
//}

public enum RenderDirection {
    case top
    case topLeft
    case topRight
    case bottom
    case bottomLeft
    case bottomRight
    case left
    case right
}

public extension CALayer {
    func renderGradient(from: RenderDirection, to: RenderDirection, colors: [UIColor]) {
        
        var cgColorsArray: [CGColor] = Array()
        
        for color in colors {
            cgColorsArray.append(color.cgColor)
        }
        
        // 初始化漸層效果
        let gradientLayer = CAGradientLayer()
        gradientLayer.startPoint    = getPoint(direction: from)
        gradientLayer.endPoint      = getPoint(direction: to)
        gradientLayer.frame         = bounds
        gradientLayer.colors        = cgColorsArray
        gradientLayer.name          = "GradientEffectsLayer"
        
        // 套用漸層效果
        if sublayers != nil && sublayers![0].name == "GradientEffectsLayer" {
            replaceSublayer(sublayers![0], with: gradientLayer)
        } else {
            insertSublayer(gradientLayer, at: 0)
        }
    }
    
    func removeRenderGradient() {
        guard let sublayers = sublayers else {
            return
        }
        
        for sublayer in sublayers {
            if sublayer.name == "GradientEffectsLayer" {
                sublayer.removeFromSuperlayer()
            }
        }
    }
    
    func getPoint(direction: RenderDirection) -> CGPoint {
        switch (direction) {
        case .top:          return CGPoint(x: 0.5, y: 0.0)
        case .topLeft:      return CGPoint(x: 0.0, y: 0.0)
        case .topRight:     return CGPoint(x: 1.0, y: 0.0)
        case .bottom:       return CGPoint(x: 0.5, y: 1.0)
        case .bottomLeft:   return CGPoint(x: 0.0, y: 1.0)
        case .bottomRight:  return CGPoint(x: 1.0, y: 1.0)
        case .left:         return CGPoint(x: 0.0, y: 0.5)
        case .right:        return CGPoint(x: 1.0, y: 0.5)
        }
    }
    
    func rotation(duration: Float) {
        let animation = CABasicAnimation(keyPath: "transform.rotation.z")
        animation.toValue        = NSNumber(value: Float.pi * 2.0)  // 旋轉角度
        animation.duration       = CFTimeInterval(duration)         // 旋轉週期
        animation.isCumulative   = true                             // 旋轉累加角度
        animation.repeatCount    = .infinity                        // 旋轉次數
        animation.timingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.linear)
        
        add(animation, forKey: "rotationAnimation")
    }
}
